crates/wysiwyg/src/node_joiner.rs

```diff
@@ -0,0 +1,207 @@
+/// and then ends up in another bold tag, the final result should be a
+/// single bold tag containing all your text.
+pub(crate) struct NodeJoiner<S>
+where
+    S: UnicodeString,
+{
+    first_type: Option<String>,
+    last_type: Option<String>,
+
+    /// This is just to allow us to have the S generic param
+    phantom_data: PhantomData<S>,
+}
+
+impl<S> NodeJoiner<S>
+where
+    S: UnicodeString,
+{
+    /// Create a NodeJoiner that handles the supplied Range, using the supplied
+    /// model to look up nodes where needed.
+    pub(crate) fn from_range(dom: &Dom<S>, range: &MultipleNodesRange) -> Self {
+        // TODO: for now, to decide whether we will join nodes, we compare
+        // the types of the parents of the start and end nodes of the range.
+        //
+        // There are some ignored tests in test_characters and test_deleting
+        // that demonstrate that this is not good enough. We need to consider
+        // joining nodes all the way up the tree.
+
+        let first_type = Self::first_text_handle(dom, range)
+            .map(|h| Self::parent_container_type(dom, h));
+
+        let last_type = Self::last_text_handle(dom, range)
+            .map(|h| Self::parent_container_type(dom, h));
+
+        Self {
+            first_type,
+            last_type,
+            phantom_data: PhantomData::default(),
+        }
+    }
+
+    /// After the selection range we were given in from_range has been deleted,
+    /// join any nodes that match up across the selection.
+    pub(crate) fn join_nodes(&self, dom: &mut Dom<S>, new_pos: usize) {
+        match (self.first_type.as_ref(), self.last_type.as_ref()) {
+            (Some(first_type), Some(last_type)) => {
+                if first_type == last_type {
+                    // The selection started and ended in the same type of
+                    // node - we should join these 2 tags.
+                    Self::join_nodes_of_same_type(dom, new_pos)
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn join_nodes_of_same_type(dom: &mut Dom<S>, new_pos: usize) {
+        let new_range = dom.find_range(new_pos, new_pos);
+
+        if let Range::SameNode(new_range) = new_range {
+            // Find the text node we are in now the selection has been deleted
+            let current_text = new_range.node_handle;
+
+            // Find the container node that contains us e.g. <b>. This is
+            // what we are going to join with something else>
+            let current_container = current_text.parent_handle();
+
+            // Find the parent that contains both the <b> tags that we are
+            // going to join.
+            let parent = current_container.parent_handle();
+
+            // Look for our sibling container - e.g. the other <b>
+            // TODO: check we have a sibling instead of crashing here
+            let sibling =
+                parent.child_handle(current_container.index_in_parent() + 1);
```

Comment:
While writing the code for formatting several nodes, I added `fn next_sibling() -> DomHandle` and `prev_sibling` for `DomHandle`. Do you think it makes sense to do it there?

Reply:
Yes, I'll do this next week